# Machine Learning Engineer Nanodegree
## Supervised Learning
## Project 2: Building a Student Intervention System

Welcome to the second project of the Machine Learning Engineer Nanodegree! In this notebook, some template code has already been provided for you, and it will be your job to implement the additional functionality necessary to successfully complete this project. Sections that begin with **'Implementation'** in the header indicate that the following block of code will require additional functionality which you must provide. Instructions will be provided for each section and the specifics of the implementation are marked in the code block with a `'TODO'` statement. Please be sure to read the instructions carefully!

In addition to implementing code, there will be questions that you must answer which relate to the project and your implementation. Each section where you will answer a question is preceded by a **'Question X'** header. Carefully read each question and provide thorough answers in the following text boxes that begin with **'Answer:'**. Your project submission will be evaluated based on your answers to each of the questions and the implementation you provide.  

>**Note:** Code and Markdown cells can be executed using the **Shift + Enter** keyboard shortcut. In addition, Markdown cells can be edited by typically double-clicking the cell to enter edit mode.

### Notes to the reviewer: Important lessons learned

After implementing the suggestions made from the first reviewer I noticed many changes happening. The changes at first seemed they would be subtle but ultimately produced significant differences in my results.

My original 3 classifiers were Decision trees, Adaboost, and Gaussian Naive Bayes. All of my original answers were based on those classifers. After making the suggested changes I almost felt it necessary to change many of answers for the project. However, even though I have learned a lot from these few changes, I do not think it would be beneficial to change all of my orginal written answers such as, explaining why I chose my 3 original classifers. Instead, I will explain here what changes I made and how they affected my final model.

#### Stratify
The first change I made was stratifying the data when splitting the data into the cross validation groups (training and test sets). The stratify parameter is used to keep the percentage of labels for the target the same or very similar between the train and test sets. Before using this parameter the graduation rate for the training set was about 68% while it was only about 63% for the test set. After stratifying, the graduation rate was approximately 67% for both the training and test sets.

I really noticed the affects of this change in the scores for my untuned classifers. Originally, my adaboost classfier performed the best at about 0.782, then the Bayes classifier was second best at 0.7634. I then tuned the adaboost classifer to get a score of 0.8000.

After adding the parameter, the new best score came from the Bayes classifer at 0.7761, second was the adaboost at only 0.7328 (almost 5% drop!).

#### Benchmarks
Next I added benchmarks. The first was to see the score we would get if we predicted that everyone was to pass. This came to 0.8030. The next benchmark was using a logistic regression with a score of 0.7500. None of my classifiers performed better than predicting all students to pass (yikes!), although the Bayes did perform better than the logistic regression classifier. 

This lead me to try the SVM classifer (SVC) to see if I could obtain a better score to beat the benchmark. The untuned SVC did perform better than the benchmark. The results of the tuned SVC are at the bottom where we compare the untuned and tuned models.

#### Base_estimator
While the documentation for adaboost says that the DecionsionTreeClassifier is the default base estimator, when actually specifying it as such it gives a lower score. The only reason I added it was so that I could tune the parameter of the adaboost and decision tree in the CVSearch parameter tunes. This lead me to add both, the original adaboost classifier (without specifying the base_estimator) and to add another adaboost with the base_estimator set to a decisiontreeclassifier. I still have some digging to do to find out why there is a diffence. This taught me it is important to set the parameters I am making assumptions about, such as setting the base_estimator to DecisionTreeClassifier or knowing for certain what the default base_estimator is.

#### SVC
I did tune the SVC and compare results at the very bottom. However, I did not add the SVC to all the other questions.

### Question 1 - Classification vs. Regression
*Your goal for this project is to identify students who might need early intervention before they fail to graduate. Which type of supervised learning problem is this, classification or regression? Why?*

**Answer: **

- This is a classification problem because the output will be discrete data, a pass or fail label, which will be output by a classification algorithm. 
- Regression on the other hand handles continuous output. This would be appropriate if we were trying to predict a students specific score on the final exam based on features provided.

## Exploring the Data
Run the code cell below to load necessary Python libraries and load the student data. Note that the last column from this dataset, `'passed'`, will be our target label (whether the student graduated or didn't graduate). All other columns are features about each student.

In [86]:
# Import libraries
import numpy as np
import pandas as pd
from time import time
from sklearn.metrics import f1_score
from sklearn.cross_validation import train_test_split

# Read student data
student_data = pd.read_csv("student-data.csv")
print "Student data read successfully!"

Student data read successfully!


### Implementation: Data Exploration
Let's begin by investigating the dataset to determine how many students we have information on, and learn about the graduation rate among these students. In the code cell below, you will need to compute the following:
- The total number of students, `n_students`.
- The total number of features for each student, `n_features`.
- The number of those students who passed, `n_passed`.
- The number of those students who failed, `n_failed`.
- The graduation rate of the class, `grad_rate`, in percent (%).


In [87]:
print student_data.head()

  school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

   ...   internet romantic  famrel  freetime  goout Dalc Walc health absences  \
0  ...         no       no       4         3      4    1    1      3        6   
1  ...        yes       no       5         3      3    1    1      3        4   
2  ...        yes       no       4         3      2    2    3      3       10   
3  ...        yes      yes       3         2      2    1    1      5        2   
4  ...         no       no       4         3      2    1    2      5        4   

  passed  
0     no  
1     no  
2    yes  
3    yes  
4

In [88]:
# TODO: Calculate number of students
n_students = len(student_data)

# TODO: Calculate number of features
n_features = len(student_data.columns)-1

# TODO: Calculate passing students
n_passed = len(student_data[student_data.passed == 'yes'])

# TODO: Calculate failing students
n_failed = len(student_data[student_data.passed == 'no'])

# TODO: Calculate graduation rate
grad_rate = (n_passed/float(n_students))*100

# Print the results
print "Total number of students: {}".format(n_students)
print "Number of features: {}".format(n_features)
print "Number of students who passed: {}".format(n_passed)
print "Number of students who failed: {}".format(n_failed)
print "Graduation rate of the class: {:.2f}%".format(grad_rate)


Total number of students: 395
Number of features: 30
Number of students who passed: 265
Number of students who failed: 130
Graduation rate of the class: 67.09%


## Preparing the Data
In this section, we will prepare the data for modeling, training and testing.

### Identify feature and target columns
It is often the case that the data you obtain contains non-numeric features. This can be a problem, as most machine learning algorithms expect numeric data to perform computations with.

Run the code cell below to separate the student data into feature and target columns to see if any features are non-numeric.

In [89]:
# Extract feature columns
feature_cols = list(student_data.columns[:-1])

# Extract target column 'passed'
target_col = student_data.columns[-1] 
# Show the list of columns
print "Feature columns:\n{}".format(feature_cols)
print "\nTarget column: {}".format(target_col)

# Separate the data into feature data and target data (X_all and y_all, respectively)
X_all = student_data[feature_cols]
y_all = student_data[target_col]

# Show the feature information by printing the first five rows
print "\nFeature values:"
print X_all.head()

Feature columns:
['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu', 'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']

Target column: passed

Feature values:
  school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

    ...    higher internet  romantic  famrel  freetime goout Dalc Walc health  \
0   ...       yes       no        no       4         3     4    1    1      3   
1   ...       

### Preprocess Feature Columns

As you can see, there are several non-numeric columns that need to be converted! Many of them are simply `yes`/`no`, e.g. `internet`. These can be reasonably converted into `1`/`0` (binary) values.

Other columns, like `Mjob` and `Fjob`, have more than two values, and are known as _categorical variables_. The recommended way to handle such a column is to create as many columns as possible values (e.g. `Fjob_teacher`, `Fjob_other`, `Fjob_services`, etc.), and assign a `1` to one of them and `0` to all others.

These generated columns are sometimes called _dummy variables_, and we will use the [`pandas.get_dummies()`](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html?highlight=get_dummies#pandas.get_dummies) function to perform this transformation. Run the code cell below to perform the preprocessing routine discussed in this section.

In [90]:
def preprocess_features(X):
    ''' Preprocesses the student data and converts non-numeric binary variables into
        binary (0/1) variables. Converts categorical variables into dummy variables. '''
    
    # Initialize new output DataFrame
    output = pd.DataFrame(index = X.index)

    # Investigate each feature column for the data
    for col, col_data in X.iteritems():
        
        # If data type is non-numeric, replace all yes/no values with 1/0
        if col_data.dtype == object:
            col_data = col_data.replace(['yes', 'no'], [1, 0])

        # If data type is categorical, convert to dummy variables
        if col_data.dtype == object:
            # Example: 'school' => 'school_GP' and 'school_MS'
            col_data = pd.get_dummies(col_data, prefix = col)  
        
        # Collect the revised columns
        output = output.join(col_data)

    return output

X_all = preprocess_features(X_all)

print "Processed feature columns ({} total features):\n{}".format(len(X_all.columns), list(X_all.columns))

Processed feature columns (48 total features):
['school_GP', 'school_MS', 'sex_F', 'sex_M', 'age', 'address_R', 'address_U', 'famsize_GT3', 'famsize_LE3', 'Pstatus_A', 'Pstatus_T', 'Medu', 'Fedu', 'Mjob_at_home', 'Mjob_health', 'Mjob_other', 'Mjob_services', 'Mjob_teacher', 'Fjob_at_home', 'Fjob_health', 'Fjob_other', 'Fjob_services', 'Fjob_teacher', 'reason_course', 'reason_home', 'reason_other', 'reason_reputation', 'guardian_father', 'guardian_mother', 'guardian_other', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']


### Implementation: Training and Testing Data Split
So far, we have converted all _categorical_ features into numeric values. For the next step, we split the data (both features and corresponding labels) into training and test sets. In the following code cell below, you will need to implement the following:
- Randomly shuffle and split the data (`X_all`, `y_all`) into training and testing subsets.
  - Use 300 training points (approximately 75%) and 95 testing points (approximately 25%).
  - Set a `random_state` for the function(s) you use, if provided.
  - Store the results in `X_train`, `X_test`, `y_train`, and `y_test`.

In [91]:
# TODO: Import any additional functionality you may need here

# TODO: Set the number of training points
num_train = 300

# Set the number of testing points
num_test = X_all.shape[0] - num_train

# TODO: Shuffle and split the dataset into the number of training and testing points above
X_train, X_test, y_train, y_test = train_test_split(X_all, y_all, stratify=y_all, test_size = 0.24, random_state = 42)


# Show the results of the split
print "Training set has {} samples.".format(X_train.shape[0])
print "Testing set has {} samples.".format(X_test.shape[0])

print "Grad rate of the training set: {:.2f}%".format(100 * (y_train == 'yes').mean())
print "Grad rate of the testing set: {:.2f}%".format(100 * (y_test == 'yes').mean())

Training set has 300 samples.
Testing set has 95 samples.
Grad rate of the training set: 67.00%
Grad rate of the testing set: 67.37%


## Training and Evaluating Models
In this section, you will choose 3 supervised learning models that are appropriate for this problem and available in `scikit-learn`. You will first discuss the reasoning behind choosing these three models by considering what you know about the data and each model's strengths and weaknesses. You will then fit the model to varying sizes of training data (100 data points, 200 data points, and 300 data points) and measure the F<sub>1</sub> score. You will need to produce three tables (one for each model) that shows the training set size, training time, prediction time, F<sub>1</sub> score on the training set, and F<sub>1</sub> score on the testing set.

### Question 2 - Model Application
*List three supervised learning models that are appropriate for this problem. What are the general applications of each model? What are their strengths and weaknesses? Given what you know about the data, why did you choose these models to be applied?*

**Answer: **

1. Decision trees 
  - General
    - They are good in situations with many features that contain different types of data such as binary and numerical.
    - Classification decision trees are represented by a set of if-then rules which leads to a decision
    - To build this tree
      - The tree is built starting from the top. The first node it creates is one that has has the most information gain. That is, the one that most closely splits the data in half. To tell how much information will be gained we look at the entropy of the data. Entropy measures the homogeneity of the data. The data starts with high entropy (mixed data). We will gain the most information by minimizing the entropy as close to 0 as we can get.
      - Once we have our first node, it's corresponding data is split into subsets. 
      - New nodes will be created with the subsets. Attributes can't be reused. Once there are no more attributes to split on the algoritm is done.
  - Industry
    - Direct marketing campaign success
    - Which gas field to invest in
  - Advantages
    - handling messy data
    - ability to deal with irrelevant features by generally ignoring them and select relevant ones first
    - Disadvantages
      - often achieve a lower generalization accuracy (which is why I will pair it with boosting
  - Why I chose
    - It's good at data with many features and mixed data types. That is the type of data that we have.

2. Ensemble learning (I will specifically use Adaboost)
  - General
    - Ensemble learners are a combination of many classifiers that are trained on small sets of data.
    - They are composed of 'weak' learners. Weak learners are those that only have to perform better with an accuracy just over 1/2 (i.e. have an error rate less than 1/2).
    - With each iteration the algorithm puts more weight on the labels that are wrong. And since we are using a weak learner, there will be information gain and some reduction in error, even if it is small.
    - These weak learners are combined using a weighted average (majority vote) technique.
    - When combined like this the algorithm can produce more accurate results than a single classifier can.
  - Industry
    - Since the base algorithm that was chosen here is a decision tree, those would be the industry examples for ensemble learning as well
    - Another example for decision trees would be assessing cost effectiveness in clinical research
  - Strengths
    - It's quick (computationally efficient)
    - no difficult parameters to set
    - versatile - wide range of base learners can be used, such as decision trees
  - Weaknesses
    - susceptible to uniform noise
    - weak learner should not be too complex to avoid overfitting
    - needs to be enough data so that the weak learning requirement is satisfied
    - Why I chose
      - Decision trees are good for this type of data set (many features with mixed data). Ensemble methods, specifically I will use AdaBoost, can use decision trees as its base to improve the accuracy. 


3. Bayesian inference
  - General
    - Bayes' rule allows us to make decisions when facing uncertainty. It allows us to use prior information with our data to come up with new information that we can use to confirm our possible suspicions.
  - Industry
    - Used for email spam identification
    - Forcasting insurance loss payments
  - Strengths
    - very efficient for training
    - performs well on many different training tasks (not sure what that means)
  - Weaknesses
    - assumes strong conditional independence, example: will not predict XOR classification accurately
  - Why I chose
    - I will use a naive bayes classifier because it is computationally efficient and because we can use our prior information to classify future data.


Why I didn't choose others
- SVM: better used with binary classifiers in 2 groups
- Neural nets: no sklearn library
- k nearest neighbors: curse of dimensionality, need exponentially more data as we add features


### Setup
Run the code cell below to initialize three helper functions which you can use for training and testing the three supervised learning models you've chosen above. The functions are as follows:
- `train_classifier` - takes as input a classifier and training data and fits the classifier to the data.
- `predict_labels` - takes as input a fit classifier, features, and a target labeling and makes predictions using the F<sub>1</sub> score.
- `train_predict` - takes as input a classifier, and the training and testing data, and performs `train_clasifier` and `predict_labels`.
 - This function will report the F<sub>1</sub> score for both the training and testing data separately.

In [92]:
def train_classifier(clf, X_train, y_train):
    ''' Fits a classifier to the training data. '''
    
    # Start the clock, train the classifier, then stop the clock
    start = time()
    clf.fit(X_train, y_train)
    end = time()
    traintime = end-start
    
    # Print the results
    print "Trained model in {:.4f} seconds".format(end - start)
    return traintime
    
def predict_labels(clf, features, target):
    ''' Makes predictions using a fit classifier based on F1 score. '''
    
    # Start the clock, make predictions, then stop the clock
    start = time()
    y_pred = clf.predict(features)
    end = time()
    predtime = end-start
    
    # Print and return results
    print "Made predictions in {:.4f} seconds.".format(end - start)
    thescore = f1_score(target.values, y_pred, pos_label='yes')
    return predtime, thescore


def train_predict(clf, X_train, y_train, X_test, y_test):
    ''' Train and predict using a classifer based on F1 score. '''
    
    # Indicate the classifier and the training set size
    print "Training a {} using a training set size of {}. . .".format(clf.__class__.__name__, len(X_train))
    clfname = clf.__class__.__name__
    setsize = len(X_train)
    
    # Train the classifier
    traintime = train_classifier(clf, X_train, y_train)
    traintime
    # Print the results of prediction for both training and testing
    
    traintimeDoNotUse, trainscore = predict_labels(clf, X_train, y_train)
    testtime, testscore = predict_labels(clf, X_test, y_test)
    
    print "F1 score for training set: {:.4f}.".format(trainscore)
    print "F1 score for test set: {:.4f}.".format(testscore)
    print "\n"
    
    return clfname, setsize, traintime, trainscore, testtime, testscore

### Implementation: Model Performance Metrics
With the predefined functions above, you will now import the three supervised learning models of your choice and run the `train_predict` function for each one. Remember that you will need to train and predict on each classifier for three different training set sizes: 100, 200, and 300. Hence, you should expect to have 9 different outputs below — 3 for each model using the varying training set sizes. In the following code cell, you will need to implement the following:
- Import the three supervised learning models you've discussed in the previous section.
- Initialize the three models and store them in `clf_A`, `clf_B`, and `clf_C`.
 - Use a `random_state` for each model you use, if provided.
 - **Note:** Use the default settings for each model — you will tune one specific model in a later section.
- Create the different training set sizes to be used to train each model.
 - *Do not reshuffle and resplit the data! The new training points should be drawn from `X_train` and `y_train`.*
- Fit each model with each training set size and make predictions on the test set (9 in total).  
**Note:** Three tables are provided after the following code cell which can be used to store your results.

In [93]:
# Logistic regression benchmark
from sklearn import linear_model
logreg = linear_model.LogisticRegression()
logreg.fit(X_train, y_train)
logpred = logreg.predict(X_test)

print 'BENCHMARKS:\n'

print "F1 score for all 'yes' on students: {:.4f}".format(
    f1_score(y_true = ['yes']*n_passed + ['no']*n_failed, y_pred = ['yes']*n_students, pos_label='yes', average='binary'))

print 'Logistic regression f1 score: {:.4f}'.format(f1_score(y_test, logpred, pos_label='yes'))


BENCHMARKS:

F1 score for all 'yes' on students: 0.8030
Logistic regression f1 score: 0.7500


In [96]:
# TODO: Import the three supervised learning models from sklearn

# classifier A
from sklearn.tree import DecisionTreeClassifier
# classifier B
from sklearn.ensemble import AdaBoostClassifier
# classifier C
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC

# TODO: Initialize the three models
clf_A = DecisionTreeClassifier(random_state=42)
clf_B = AdaBoostClassifier()
clf_C = AdaBoostClassifier(base_estimator=DecisionTreeClassifier())
clf_D = GaussianNB()
clf_E = SVC(random_state=42)

# TODO: Set up the training set sizes
X_train_100 = X_train[:100]
y_train_100 = y_train[:100]

X_train_200 = X_train[:200]
y_train_200 = y_train[:200]

X_train_300 = X_train[:300]
y_train_300 = y_train[:300]

# TODO: Execute the 'train_predict' function for each classifier and each training set size
colnames = ['Classifier', 'Training Set Size', 'Training Time', 'Prediction Time (test)', 
                                  'F1 Score (train)', 'F1 Score (test)']
dftable = pd.DataFrame(columns = colnames)
clflist = [clf_A, clf_B, clf_C, clf_D, clf_E]
setlist = [100,200,300]
for clffnc in range(len(clflist)):
    clf = clflist[clffnc]
    for dataset in setlist:
        X_trainer = X_train[:dataset]
        y_trainer = y_train[:dataset]
        X_tester = X_test[:dataset]
        y_tester = y_test[:dataset]
        clfname, setsize, traintime, trainscore, testtime, testscore = train_predict(clf, X_trainer, y_trainer, X_tester, y_tester)
        s1 = pd.Series([clfname, int(setsize), round(traintime,4), round(testtime,4), round(trainscore,4), round(testscore,4)], index=[colnames])
        dftable = dftable.append(s1, ignore_index=True)

        
#dftable.style.apply(dftable[dftable.loc[3]], color='yellow', axis=1)
dftable.style
                       

Training a DecisionTreeClassifier using a training set size of 100. . .
Trained model in 0.0020 seconds
Made predictions in 0.0010 seconds.
Made predictions in 0.0010 seconds.
F1 score for training set: 1.0000.
F1 score for test set: 0.6452.


Training a DecisionTreeClassifier using a training set size of 200. . .
Trained model in 0.0040 seconds
Made predictions in 0.0010 seconds.
Made predictions in 0.0000 seconds.
F1 score for training set: 1.0000.
F1 score for test set: 0.7258.


Training a DecisionTreeClassifier using a training set size of 300. . .
Trained model in 0.0050 seconds
Made predictions in 0.0000 seconds.
Made predictions in 0.0000 seconds.
F1 score for training set: 1.0000.
F1 score for test set: 0.6838.


Training a AdaBoostClassifier using a training set size of 100. . .
Trained model in 0.2470 seconds
Made predictions in 0.0140 seconds.
Made predictions in 0.0130 seconds.
F1 score for training set: 0.9778.
F1 score for test set: 0.6880.


Training a AdaBoostClassifie

### Tabular Results
Edit the cell below to see how a table can be designed in [Markdown](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#tables). You can record your results from above in the tables provided.

Please see my table above.

** Classifer 1 - Please see my table above**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |                         |                        |                  |                 |
| 200               |        EXAMPLE          |                        |                  |                 |
| 300               |                         |                        |                  |    EXAMPLE      |

** Classifer 2 - ?**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |                         |                        |                  |                 |
| 200               |     EXAMPLE             |                        |                  |                 |
| 300               |                         |                        |                  |     EXAMPLE     |

** Classifer 3 - ?**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |                         |                        |                  |                 |
| 200               |                         |                        |                  |                 |
| 300               |                         |                        |                  |                 |

## Choosing the Best Model
In this final section, you will choose from the three supervised learning models the *best* model to use on the student data. You will then perform a grid search optimization for the model over the entire training set (`X_train` and `y_train`) by tuning at least one parameter to improve upon the untuned model's F<sub>1</sub> score. 

### Question 3 - Chosing the Best Model
*Based on the experiments you performed earlier, in one to two paragraphs, explain to the board of supervisors what single model you chose as the best model. Which model is generally the most appropriate based on the available data, limited resources, cost, and performance?*

**Answer: **
- The adaboost classifier is the best model for predicting who is likely to pass or fail. Based on the available data it is the classifer that scores the highest and will most accurately make the predictions we are looking for. While the adaboost was the least efficient in terms of computing resources, the value it creates with a significantly better performance outweights the added computational resources. However, it is important to point out that while the adaboost classifier has the highest cost in regards to computing resources, the actual resouces it needs are still negligable. This is all demonstrated in table1: Classifiers with attributes

### Question 4 - Model in Layman's Terms
*In one to two paragraphs, explain to the board of directors in layman's terms how the final model chosen is supposed to work. For example if you've chosen to use a decision tree or a support vector machine, how does the model go about making a prediction?*

**Answer: **

Adaboost is an algorithm that is very similar to a class of students working on a project. The students will work on the project one-at-a-time and each will have a different part, or section, of the project to work on. After each student is done, the teacher will grade how well the previous student did. After grading, the next student is going to work on their part of the project. However there is a catch, while it might not be completly fair, the next student is going to have to work on all the areas that the first student received a poor grade (in this example the poor grades are associated with the hardest part of the project) and very little in other 'easier' areas of the project.

Each student will make at least some positive advancement on the project even if it is just a little bit. This continues until the project is complete. The result is a project with a better grade than any single student could have done by themselves.

This is how the specific algorithm works. This is considered a supervised learning algorithm. In supervised learning, we use given data, in this case it is data indicating whether a student passed or failed. With this data we will train our algorithm to find the best model that will predict future outcomes with the most accuracy. It will then use this model to predict whether or not current students will pass or fail.


### Implementation: Model Tuning
Fine tune the chosen model. Use grid search (`GridSearchCV`) with at least one important parameter tuned with at least 3 different values. You will need to use the entire training set for this. In the code cell below, you will need to implement the following:
- Import [`sklearn.grid_search.gridSearchCV`](http://scikit-learn.org/stable/modules/generated/sklearn.grid_search.GridSearchCV.html) and [`sklearn.metrics.make_scorer`](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html).
- Create a dictionary of parameters you wish to tune for the chosen model.
 - Example: `parameters = {'parameter' : [list of values]}`.
- Initialize the classifier you've chosen and store it in `clf`.
- Create the F<sub>1</sub> scoring function using `make_scorer` and store it in `f1_scorer`.
 - Set the `pos_label` parameter to the correct value!
- Perform grid search on the classifier `clf` using `f1_scorer` as the scoring method, and store it in `grid_obj`.
- Fit the grid search object to the training data (`X_train`, `y_train`), and store it in `grid_obj`.

In [45]:
# TODO: Import 'GridSearchCV' and 'make_scorer'
from sklearn.grid_search import GridSearchCV
from sklearn.metrics import make_scorer
# TODO: Create the parameters list you wish to tune
parameters = {'n_estimators':[5,10,20,50,100,500,1000], 
              'learning_rate':[0.001, 0.01,0.1,0.3, 1.,1.5,3.],
              'base_estimator__min_samples_split' : np.arange(2, 8, 2),
              'base_estimator__max_depth' : np.arange(2, 6, 2)}

# TODO: Initialize the classifier
clf = AdaBoostClassifier(base_estimator=DecisionTreeClassifier()) # default classifier is DecicionTreeClassifier()

# TODO: Make an f1 scoring function using 'make_scorer' 
f1_scorer = make_scorer(f1_score, pos_label='yes')

# TODO: Perform grid search on the classifier using the f1_scorer as the scoring method
grid_obj = GridSearchCV(estimator = clf, param_grid = parameters, scoring = f1_scorer, cv=4)


# TODO: Fit the grid search object to the training data and find the optimal parameters
grid_obj = grid_obj.fit(X_train, y_train)

# Get the estimator
clf = grid_obj.best_estimator_

# Report the final F1 score for training and testing after parameter tuning
predtime, trainscore = predict_labels(clf, X_train, y_train)
predtime, testscore = predict_labels(clf, X_test, y_test)
print "Tuned model has a training F1 score of {:.4f}.".format(trainscore)
print "Tuned model has a testing F1 score of {:.4f}.".format(testscore)

#print "Tuned model has a training F1 score of {:.4f}.".format(predict_labels(clf, X_train, y_train))
#print "Tuned model has a testing F1 score of {:.4f}.".format(predict_labels(clf, X_test, y_test))

Made predictions in 0.0010 seconds.
Made predictions in 0.0010 seconds.
Tuned model has a training F1 score of 0.8246.
Tuned model has a testing F1 score of 0.7681.


In [95]:
print grid_obj.best_params_, '\n'

{'n_estimators': 5, 'base_estimator__min_samples_split': 2, 'base_estimator__max_depth': 2, 'learning_rate': 0.01} 



In [109]:
## SVC tune
# TODO: Import 'GridSearchCV' and 'make_scorer'
from sklearn.grid_search import GridSearchCV
from sklearn.metrics import make_scorer
# TODO: Create the parameters list you wish to tune
parameters = {'C':[0.01,0.1,1.0,2.0,3.0,5.0,10.0,100.00,1000.0]
              }

# TODO: Initialize the classifier
clf = SVC(random_state=42) 

# TODO: Make an f1 scoring function using 'make_scorer' 
f1_scorer = make_scorer(f1_score, pos_label='yes')

# TODO: Perform grid search on the classifier using the f1_scorer as the scoring method
grid_obj = GridSearchCV(estimator = clf, param_grid = parameters, scoring = f1_scorer, cv=4)


# TODO: Fit the grid search object to the training data and find the optimal parameters
grid_obj = grid_obj.fit(X_train, y_train)

# Get the estimator
clf = grid_obj.best_estimator_

# Report the final F1 score for training and testing after parameter tuning
predtime, trainscore = predict_labels(clf, X_train, y_train)
predtime, testscore = predict_labels(clf, X_test, y_test)
print "Tuned model has a training F1 score of {:.4f}.".format(trainscore)
print "Tuned model has a testing F1 score of {:.4f}.".format(testscore)

#print "Tuned model has a training F1 score of {:.4f}.".format(predict_labels(clf, X_train, y_train))
#print "Tuned model has a testing F1 score of {:.4f}.".format(predict_labels(clf, X_test, y_test))

Made predictions in 0.0110 seconds.
Made predictions in 0.0040 seconds.
Tuned model has a training F1 score of 0.9157.
Tuned model has a testing F1 score of 0.8054.


In [110]:
print grid_obj.best_params_, '\n'

{'C': 2.0} 



### Question 5 - Final F<sub>1</sub> Score
*What is the final model's F<sub>1</sub> score for training and testing? How does that score compare to the untuned model?*

**Answer: **

##### For the original adaboost tune
- The final model's F<sub>1</sub> training score was: 0.8246
  - This score was lower than the untuned model's score of 1.0
- The final model's F<sub>1</sub> testing score was: 0.7681
  - This score was higher than the untuned model's score of 0.7167
- Having a lower score for the training score is not bad. The tuend model did not overfit as much as the untuned model did. This is a good thing because your training model more closely resembles what happens with the test data. Most important, the test score was higher than the untuned model.


##### For the SVC tune
- The final model's F<sub>1</sub> training score was: 0.9157
  - This score was lower than the untuned model's score of 0.8664
- The final model's F<sub>1</sub> testing score was: 0.8054
  - This score was higher than the untuned model's score of 0.8052
- The tuning barely made an improvement 

> **Note**: Once you have completed all of the code implementations and successfully answered each question above, you may finalize your work by exporting the iPython Notebook as an HTML document. You can do this by using the menu above and navigating to  
**File -> Download as -> HTML (.html)**. Include the finished document along with this notebook as your submission.